File: src/main/java/pl/tsimur/jpatest/exception/GlobalExceptionHandler.java
package pl.tsimur.jpatest.exception;

import lombok.extern.slf4j.Slf4j;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestControllerAdvice;
import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler;

import javax.persistence.EntityNotFoundException;

import static org.springframework.http.HttpStatus.*;

@Slf4j
@RestControllerAdvice
public class GlobalExceptionHandler extends ResponseEntityExceptionHandler {

    @ExceptionHandler(value = {Exception.class})
    public String unknownExceptionHandler(Exception ex) {
        log.error("🐥 UNKNOWN: " + ex.getMessage());
        return ex.getMessage();
    }

    @ResponseStatus(NOT_FOUND)
    @ExceptionHandler(value = {EntityNotFoundException.class, NotFoundException.class})
    public String notFoundExceptionHandler(Exception ex) {
        log.error("⛑ NOT FOUND: " + ex.getMessage());
        return ex.getMessage();
    }

    @ResponseStatus(PAYMENT_REQUIRED)
    @ExceptionHandler(value = {NotEnoughMoneyException.class})
    public String notEnoughMoneyExceptionHandler(Exception ex) {
        log.error("💰 NOT ENOUGH: " + ex.getMessage());
        return ex.getMessage();
    }

    @ResponseStatus(BAD_REQUEST)
    @ExceptionHandler(value = {InvalidArgumentException.class})
    public String invalidArgumentExceptionHandler(Exception ex) {
        log.error("🤨 INVALID: " + ex.getMessage());
        return ex.getMessage();
    }
}

File: src/main/kotlin/pl/timur/jpatest/exception/GlobalExceptionHandler.kt
package pl.timur.jpatest.exception

import org.slf4j.LoggerFactory
import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.RestControllerAdvice
import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler
import javax.persistence.EntityNotFoundException

@RestControllerAdvice
class GlobalExceptionHandler : ResponseEntityExceptionHandler() {

    companion object {
        private val log = LoggerFactory.getLogger(this::class.java)
    }

    @ExceptionHandler(value = [Exception::class])
    fun unknownExceptionHandler(ex: Exception): String? {
        log.error("🐥 UNKNOWN: " + ex.message)
        return ex.message
    }

    @ResponseStatus(HttpStatus.NOT_FOUND)
    @ExceptionHandler(value = [EntityNotFoundException::class, NotFoundException::class])
    fun notFoundExceptionHandler(ex: Exception): String? {
        log.error("⛑ NOT FOUND: " + ex.message)
        return ex.message
    }

    @ResponseStatus(HttpStatus.PAYMENT_REQUIRED)
    @ExceptionHandler(value = [NotEnoughMoneyException::class])
    fun notEnoughMoneyExceptionHandler(ex: Exception): String? {
        log.error("💰 NOT ENOUGH: " + ex.message)
        return ex.message
    }

    @ResponseStatus(HttpStatus.BAD_REQUEST)
    @ExceptionHandler(value = [InvalidArgumentException::class])
    fun invalidArgumentExceptionHandler(ex: Exception): String? {
        log.error("🤨 INVALID: " + ex.message)
        return ex.message
    }
}